src/ui/wxWidgets/wxUtilities.cpp

```diff
@@ -105,38 +113,49 @@ void CloseChildWindowRecursively(wxTopLevelWindow* win, wxTopLevelWindow* top)
   }
 }
 
-void HideWindowRecursively(wxTopLevelWindow* win)
+void HideWindowRecursively(wxTopLevelWindow* win, wxWindowList& hiddenWindows)
 {
-  if (!win)
+  if (!win || !win->IsShown())
     return;
   wxWindowList& children = win->GetChildren();
+  pws_os::Trace(L"HideWindowRecursively: %ls\n", ToStr(win->GetTitle()));
   for(wxWindowList::iterator itr = children.begin(); itr != children.end(); ++itr) {
+    auto ww = wxDynamicCast(*itr, wxTopLevelWindow);
+    if (ww) {
+      pws_os::Trace(L"HideWindowRecursively2: %ls\n", ToStr(ww->GetTitle()));
+    }
     if ((*itr)->IsTopLevel() && (*itr)->IsShown()) {
-      HideWindowRecursively(wxDynamicCast(*itr, wxTopLevelWindow));
+      HideWindowRecursively(wxDynamicCast(*itr, wxTopLevelWindow), hiddenWindows);
     }
   }
 
   //Don't call Hide() here, which just calls Show(false), which is overridden in
   //derived classes, and wxDialog actually cancels the modal loop and closes the window
   win->wxWindow::Show(false);
+  // we need to remember hidden windows to show later only them
+  hiddenWindows.push_front(win);
 }
 
-void ShowWindowRecursively(wxTopLevelWindow* win)
+void ShowWindowRecursively(wxWindowList& hiddenWindows)
 {
-  if (!win)
-    return;
-  // Handle at first to ensure Show() in the reverse order of Hide()'ing
-  wxASSERT(win->IsTopLevel());
-  win->wxWindow::Show(true);
-  win->Raise();
-  win->Update();
-  
-  wxWindowList& children = win->GetChildren();
-  for(wxWindowList::iterator itr = children.begin(); itr != children.end(); ++itr) {
-    if ((*itr)->IsTopLevel()) {
-      ShowWindowRecursively(wxDynamicCast(*itr, wxTopLevelWindow));
-    }
+  for(wxWindowList::iterator itr = hiddenWindows.begin(); itr != hiddenWindows.end(); ++itr) {
+    wxWindow* win = (*itr);
+    win->wxWindow::Show(true);
+    win->Raise();
+    win->Update();
   }
+  hiddenWindows.clear();
+}
+
+wxTopLevelWindow* GetNearestTlw(wxWindow* win) {
```

Comment:
`GetNearestTopLevelWindow`?

Actually, the implementation seems identical to the other method. Remove?

Reply:
Will rename in final version.